### DISCUSSION TITLE
How to update data when only query changes? watchQuery is no longer supported..

### DISCUSSION THREAD

**Discussion by johannesstricker** [nuxt/nuxt#16113] [Questions] [answered]
When `useFetch` is using query parameters from `useRoute`, it doesn't get updated when the route path remains unchanged, but the query parameters change. For example:
```js
// pages/search.vue

<script setup>
const route = useRoute();
const { data } = useFetch('/path/to/api', { params: { q: route.query.q } });

// this changes the query params, but the data remains stale
const router = useRouter();
const submitSearch = (text) => router.push({ 
  path: '/search', 
  query: { q: text },
});
</script setup>
```

In Nuxt2 you could use [watchQuery](https://nuxtjs.org/docs/components-glossary/watchquery/) for this, but it no longer seems to be supported. I could only fix this using a combination of `useAsyncData` and `watch`. 

```js
const route = useRoute();
const reactiveQuery = computed(() => route.query.q);
const { data, refresh } = useAsyncData('key', $fetch('/path/to/api', params: {
  q: reactiveQuery.value,
});
watch(reactiveQuery, refresh);
```

Is there a better way to get this working?

**Comment by wobsoriano** [comment, marked as answer]
You are doing it right

**Reply by johannesstricker** [reply]
Okay, unfortunate that it is that complicated. Thank you.

**Comment by ironytr** [comment]
> When `useFetch` is using query parameters from `useRoute`, it doesn't get updated when the route path remains unchanged, but the query parameters change. For example:
> 
> ```js
> // pages/search.vue
> 
> <script setup>
> const route = useRoute();
> const { data } = useFetch('/path/to/api', { params: { q: route.query.q } });
> 
> // this changes the query params, but the data remains stale
> const router = useRouter();
> const submitSearch = (text) => router.push({ 
>   path: '/search', 
>   query: { q: text },
> });
> </script setup>
> ```
> 
> In Nuxt2 you could use [watchQuery](https://nuxtjs.org/docs/components-glossary/watchquery/) for this, but it no longer seems to be supported. I could only fix this using a combination of `useAsyncData` and `watch`.
> 
> ```js
> const route = useRoute();
> const reactiveQuery = computed(() => route.query.q);
> const { data, refresh } = useAsyncData('key', $fetch('/path/to/api', params: {
>   q: reactiveQuery.value,
> });
> watch(reactiveQuery, refresh);
> ```
> 
> Is there a better way to get this working?

Thank you for that snippet, if there new way to published for this would you reply to this topic so noobs like me can use that .

**Reply by johannesstricker** [reply]
I found out that you can also assign a `key` to the slot that the router renders into, e.g.

```
// layouts/default.vue

<template>
  <main>
    <slot :key="route.fullPath" />
  </main>
</template>
```

This will cause the slot to re-render whenever the route (including query parameters) changes. This might have some unwanted side-effects though 🤷‍♂️

**Comment by papoola** [comment]
I found a better shorter solution to this problem, just define a key based on something that is different ( e.g. query parameter ):

```js
const { data } = useFetch('/path/to/api', {
  key: '/path/to/api/' + route.query.q
  params: { q: route.query.q }
});
```

This will tell nuxt that this is a different API call, hence update occurs

**Comment by TheAlexLichter** [comment]
The minimal version needed is using a computed property for the params. See [this stackblitz](https://stackblitz.com/edit/nuxtframeworkdiscussions2568?file=app.vue).

```ts
const route = useRoute();
const postId = computed(() => route.query.postId);
const { data } = useFetch('https://jsonplaceholder.typicode.com/comments', {
  params: { postId },
});
```
